net/src/lib.rs

```diff
@@ -0,0 +1,373 @@
+//! holochain_net is a library that defines an abstract networking layer for
+//! different network transports and implements a PeerStore for mapping and
+//! managing the topology of transport layers with regard to relay's e.g. for NAT
+
+#[macro_use]
+extern crate failure;
+
+use failure::Error;
+use std::sync::{Arc, Mutex};
+
+pub mod error;
+
+pub type SerializedAddress = Vec<u8>;
+pub type TransportAddress = Vec<u8>;
+pub type SerializedMessage = Vec<u8>;
+
+/// this closure type will get called when the send completes and the parameter will be the response message (or error)
+type SendResponseClosure = Box<FnMut(Result<SerializedMessage, Error>) -> Option<Error> + Send>;
+
+/// this closure type gets called when a new message arrives, you can respond with a Message or an error.
+type ReceiveClosure =
+    Box<FnMut(&SerializedAddress, &SerializedMessage) -> Result<SerializedMessage, Error> + Send>;
+
+pub trait Node {
+    fn get_address(&self) -> &SerializedAddress;
+    fn get_transport_address(&self) -> TransportAddress;
+}
+
+pub struct Handler {
+    pub handler: Option<ReceiveClosure>,
+}
+
+impl Handler {
+    fn handle(
+        &mut self,
+        from: &SerializedAddress,
+        message: &SerializedMessage,
+    ) -> Result<SerializedMessage, Error> {
+        match self.handler {
+            None => bail!("fish"),
+            Some(ref mut handler) => (handler)(from, message),
+        }
+    }
+}
+
+pub trait Transport {
+    /** initialize the transport
+     * this might be used for example in a TCP based transport to specify a listening port for
+     * for the transport.
+     */
+    fn initialize(&mut self, config: Option<String>) -> Result<(), Error>;
```

Comment:
Shouldn't initialize return the initial Transport? In mean, you don't want something to be initialized twice or more often.

Reply:
You can't make initialize return an object that doesn't exist yet, and I don't want this to turn into a trait object, because that's what caused all the problems first time around.  Perhaps it could return self for chaining, but it some Transports I could imagine that re-initializing is semantically ok.  We would need to declare an expected behavior if it's not